File: packages/telemetry/src/__tests__/evaluate-stage-gates.test.ts
import { describe, expect, it, vi } from 'vitest';
import { evaluateStageGates } from '../evaluate/evaluate-stage-gates.js';
import * as evaluateGatePolicyModule from '../evaluate/evaluate-gate-policy.js';

const runContext = {
	flagKey: 'demo-feature-flag',
	treatmentVariationId: 'treatment-var',
	controlVariationId: 'control-var',
};

describe('evaluateStageGates', () => {
	it('passes when all policies pass', async () => {
		const evaluateSpy = vi
			.spyOn(evaluateGatePolicyModule, 'evaluateGatePolicy')
			.mockImplementation(async (_client, policy) => ({
				verdict: 'pass',
				metricType: policy.metricType,
				threshold: policy.threshold,
				metadata: {},
			}));

		const result = await evaluateStageGates(
			{ config: { baseUrl: 'http://localhost:9090' }, queryInstant: vi.fn() },
			[
				{ metricType: 'error_rate', threshold: 0.01, serviceName: 'demo-service' },
				{ metricType: 'latency_p95', threshold: 500, serviceName: 'demo-service' },
			],
			runContext,
		);

		expect(result.verdict).toBe('pass');
		expect(result.results).toHaveLength(2);
		evaluateSpy.mockRestore();
	});

	it('fails when any policy fails', async () => {
		const evaluateSpy = vi
			.spyOn(evaluateGatePolicyModule, 'evaluateGatePolicy')
			.mockImplementation(async (_client, policy) => ({
				verdict: policy.metricType === 'latency_p95' ? 'fail' : 'pass',
				metricType: policy.metricType,
				threshold: policy.threshold,
				metadata:
					policy.metricType === 'latency_p95'
						? { reason: 'threshold_exceeded' }
						: {},
			}));

		const result = await evaluateStageGates(
			{ config: { baseUrl: 'http://localhost:9090' }, queryInstant: vi.fn() },
			[
				{ metricType: 'error_rate', threshold: 0.01, serviceName: 'demo-service' },
				{ metricType: 'latency_p95', threshold: 500, serviceName: 'demo-service' },
			],
			runContext,
		);

		expect(result.verdict).toBe('fail');
		evaluateSpy.mockRestore();
	});
});

File: kotlin/modules/telemetry/src/test/kotlin/com/ffpromo/telemetry/evaluate/EvaluateStageGatesTest.kt
package com.ffpromo.telemetry.evaluate

import com.ffpromo.contracts.GatePolicyInput
import com.ffpromo.contracts.GateRunContext
import com.ffpromo.contracts.GateVerdictResult
import com.ffpromo.contracts.MetricType
import com.ffpromo.telemetry.support.QueuePrometheusClient
import com.ffpromo.telemetry.support.vector
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class EvaluateStageGatesTest {
    private val runContext = GateRunContext(
        flagKey = "demo-feature-flag",
        treatmentVariationId = "treatment-var",
        controlVariationId = "control-var",
    )

    @Test
    fun `passes when all policies pass`() = runTest {
        val client = QueuePrometheusClient(
            mutableListOf(
                vector("0.02"), vector("0.01"), vector("200"), vector("200"),
                vector("300"), vector("300"), vector("200"), vector("200"),
            ),
        )
        val result = evaluateStageGates(
            client,
            listOf(
                GatePolicyInput(MetricType.error_rate, 0.01, "demo-service", minSampleSize = 100),
                GatePolicyInput(MetricType.latency_p95, 500.0, "demo-service", minSampleSize = 100),
            ),
            runContext,
        )
        assertEquals(GateVerdictResult.pass, result.verdict)
        assertEquals(2, result.results.size)
    }

    @Test
    fun `fails when any policy fails`() = runTest {
        val client = QueuePrometheusClient(
            mutableListOf(
                vector("0.02"), vector("0.01"), vector("200"), vector("200"),
                vector("1000"), vector("300"), vector("200"), vector("200"),
            ),
        )
        val result = evaluateStageGates(
            client,
            listOf(
                GatePolicyInput(MetricType.error_rate, 0.01, "demo-service", minSampleSize = 100),
                GatePolicyInput(MetricType.latency_p95, 500.0, "demo-service", minSampleSize = 100),
            ),
            runContext,
        )
        assertEquals(GateVerdictResult.fail, result.verdict)
    }
}
